### PR TITLE
Python: Refactor schema compatibility check

### PR DESCRIPTION
**Pull request by Fokko**
This splits out checking schema compatibility from `to_requested_schema` to `check_schema_compatibility` and this way we can create empty structs in `to_requested_schema` because we already know that everything is compatible.

### CODE REVIEW

File: python/pyiceberg/io/pyarrow.py [hunk 1]
```diff
@@ -518,15 +519,14 @@ def project_table(
         ).to_table(filter=pyarrow_filter)
 
         tables.append(to_requested_schema(projected_schema, file_project_schema, arrow_table))
-
```
Comment (rdblue) [inline python/pyiceberg/io/pyarrow.py:521]: Nit: unnecessary whitespace change.

File: python/pyiceberg/schema.py [hunk 2]
```diff
@@ -1350,3 +1349,98 @@ def _(file_type: DecimalType, read_type: IcebergType) -> IcebergType:
...
+
+    def schema_partner(self, partner: Optional[Union[Schema, IcebergType]]) -> Optional[Union[Schema, IcebergType]]:
+        return partner
+
+    def field_partner(
+        self, partner_struct: Optional[Union[Schema, IcebergType]], field_id: int, _: str
+    ) -> Optional[Union[Schema, IcebergType]]:
+        if partner_struct is not None:
+            try:
+                return self.partner_schema.find_field(field_id).field_type
```
Comment (rdblue) [inline python/pyiceberg/schema.py:1434]: Visitors should not use `find` methods. Can you rewrite this to use `struct.field(field_id)` instead?

File: python/pyiceberg/schema.py [hunk 3]
```diff
@@ -1350,3 +1349,98 @@ def _(file_type: DecimalType, read_type: IcebergType) -> IcebergType:
...
+    def primitive(
+        self, primitive: PrimitiveType, primitive_partner: Optional[Union[Schema, IcebergType]]
+    ) -> Union[Schema, IcebergType]:
+        if primitive_partner and primitive_partner != primitive:
+            return promote(primitive_partner, primitive)
+        else:
+            return primitive
+
+
+class _PartnerSchemaAccessor(PartnerAccessor[Union[Schema, IcebergType]]):
```
Comment (rdblue) [inline python/pyiceberg/schema.py:1420]: Can we unify this with `SchemaPartnerAccessor` that is used in the resolver? That one will raise exceptions, but we could either have a flag to disable exceptions or we could update the tests to not expect them. I don't actually think the exceptions are very useful, since we always expect the structure to match.

File: python/tests/io/test_pyarrow.py [hunk 4]
```diff
@@ -1014,7 +1014,7 @@ def test_projection_nested_new_field(file_struct: str) -> None:
     result_table = project(schema, [file_struct])
 
     for actual, expected in zip(result_table.columns[0], [None, None, None]):
-        assert actual.as_py() == {"null": expected}
+        assert actual.as_py() == expected
```
Comment (rdblue) [inline python/tests/io/test_pyarrow.py:1017]: I think this change is incorrect. There is a non-null location in every row so there should be a non-null struct for each result. The null/None value is the "null" column that doesn't exist.

File: python/pyiceberg/io/pyarrow.py [hunk 5]
```diff
@@ -561,15 +557,13 @@ def struct(
         fields: List[pa.Field] = []
         for field, field_array in zip(struct.fields, field_results):
             if field_array is not None:
-                array = self.cast_if_needed(field, field_array)
+                array = self._cast_if_needed(field, field_array)
                 field_arrays.append(array)
                 fields.append(pa.field(field.name, array.type, field.optional))
-            elif field.optional:
```
Comment (Fokko) [inline python/pyiceberg/io/pyarrow.py:567]: This check is done in the `check_schema_compatibility`

File: python/pyiceberg/schema.py (hunk 3 above)
Comment (Fokko) [reply python/pyiceberg/schema.py:1420]: Yes, we can and I like that very much

**Comment by Fokko** [comment]
I'm not sure if this is an improvement
